### ISSUE TITLE
Line ending confusion in generated files of Xtext project?

### ISSUE THREAD

**Issue by Bananeweizen** [eclipse/xtext#2778]
When I create a new Oomph setup for the Xtext project on my Windows machine, then Oomph automatically sets autocrlf=true, even though there is no such attribute in the *.setup file:
![grafik](https://github.com/eclipse/xtext/assets/406876/b9752952-b986-4ead-bf3a-0d649ee19391)

That in turn leads to all the Xtend generated Java files having CRLF in my local index (while the Github repo uses LF). But during build and re-generation all these files get LF only (which fits to the projects having the Unix line ending setting in their project properties). Therefore the IDE has 800+ changed files after a restart, almost all of them with newline changes only, looking like this:
![grafik](https://github.com/eclipse/xtext/assets/406876/57aa8d36-90dd-4ab6-9ebb-3d03f92e1c0e)

I do use a personal Oomph setup, but I'm not aware of any setting that would cause Oomph to add the autocrlf just for me, so I suspect this should be broken similarly for everyone on Windows. If that is true, would it be useful to explicitly set autocrlf to another value in the *.setup or alternatively to set `*.java text eol=lf` in a `.gitattributes` file eventually? I've used the gitattributes approach locally, reset hard, and since then my staging area has only 24 changes files, where most contain actual non whitespaces changes (which would be another thing to investigate later).

**Comment by cdietrich** [comment]
this issue is new to me. 
i am only aware on windows changes on fresh workspaces caused by 
https://github.com/eclipse/xtext/issues/2293

unfortunately with company change i have no suiteable windows machine anymore to have a look.

**Comment by Bananeweizen** [comment]
The Windows changes that you linked are in the above mentioned remaining 24 changed files after I applied the gitattributes. I can try to have a colleague set up another Xtext installation on his machine next week to verify he sees the same.
My initial thought was that this may have been caused by the monorepo change some weeks/months ago, but I have not been able to pinpoint to any changes in the .setup or in the archived repos that would support that claim.

**Comment by cdietrich** [comment]
maybe a default on oomph or egit side has changed?

**Comment by pbrossel** [comment]
Hi @Bananeweizen,
can you tell me what changes you have made to temp fix the issue ?
I did an initial oomph setup on windows and have exactly the same problem as you had.
>...
I've used the gitattributes approach locally, reset hard, and since then my staging area has only 24 changes files, where most contain actual non whitespaces changes (which would be another thing to investigate later).

**Comment by LorenzoBettini** [comment]
I've just tried the Oomph setup in Windows as well. I mostly use Linux and sometimes macOS, while I rarely used Windows.

I'm experiencing the same problem as well!

We are aware of a few Xtend generated files with Windows EOLs, which is https://github.com/eclipse/xtext/issues/2293 for which I recently proposed a fix https://github.com/eclipse/xtext/pull/3004

However, I experienced the bigger problem reported here, I seem to understand:

- after the first Oomph run, there are only a few generated Xtend files that are dirty (because of the generated "\r" in the Java code's strings)
- as soon as I restart the Eclipse workspace, Oomph insists in always re-running our task "Projects Build (Refresh, Clean, Build)" (@merks any idea why? This does not happen in Linux and macOS; why does Oomph always re-execute that task?) which then leads to tons of dirty files, not related to https://github.com/eclipse/xtext/issues/2293 but to the problem initially reported in this bug.

I seem to understand that "autocrlf=true" is the correct setting in Windows, and @HannesWell also recently provided a merged PR normalizing a few files with incorrect line endings.

Of course, I can reset hard, but this is very annoying.

@Bananeweizen says that

> But during build and re-generation all these files get LF only (which fits to the projects having the Unix line ending setting in their project properties).

So maybe we should not have that setting in our projects?

**Comment by LorenzoBettini** [comment]
A few more experiments (in Windows): the projects where this happens (again, independently from https://github.com/eclipse/xtext/issues/2293) seem to be where we have `/org.eclipse.xtend.ide/.settings/org.eclipse.core.runtime.prefs` with these contents>

```
eclipse.preferences.version=1
line.separator=\n
```

Removing the file and rebuilding fixes the problem in that project.

@cdietrich I can provide a PR removing all such files, what do you think?

**Comment by cdietrich** [comment]
unfortunately i have no idea.
i assume m2e or something else will replace this again

**Comment by LorenzoBettini** [comment]
I see those files were added here https://github.com/search?q=repo%3Aeclipse%2Fxtext+Set+UNIX-style+linebreaks&type=commits but I see no reason why they are there.

No, @cdietrich , the files you're mentioning, which are correctly regenerated/updated, are `/org.eclipse.xtend.ide/.settings/org.eclipse.core.resources.prefs`; the ones I talk about, which are not even everywhere, are `/org.eclipse.xtend.ide/.settings/org.eclipse.core.runtime.prefs` (runtime vs resources).

**Comment by cdietrich** [comment]
@szarnekow do you remember why its this way

**Comment by LorenzoBettini** [comment]
With this https://github.com/eclipse/xtext/pull/3005 on my Windows system, even a clean build does not generate all the dirty files (only seven remain due to https://github.com/eclipse/xtext/issues/2293)

**Comment by HannesWell** [comment]
Probably having `autocrlf=true`in the git config is overall the better way to keep the line endings consistent in git.

> the files you're mentioning, which are correctly regenerated/updated, are /org.eclipse.xtend.ide/.settings/org.eclipse.core.resources.prefs; the ones I talk about, which are not even everywhere, are /org.eclipse.xtend.ide/.settings/org.eclipse.core.runtime.prefs (runtime vs resources)

The resource preferences are even generated by the eclipse core (probably even the resources plugin itself). This changed a few releases ago.

Regarding m2e updating the configuration: Automatic updates can be disabled in the workspace preferences and since the last release even in the project preferences.

**Comment by LorenzoBettini** [comment]
@HannesWell I'm talking about "runtime.prefs", not "resources.prefs".

**Comment by HannesWell** [comment]
Yes. My second paragraph was just for information in case you didn't knew it already and it is of interest for you.

**Comment by szarnekow** [comment]
> @szarnekow do you remember why its this way

I can imagine that it was done to get a stable compilation result on windows and linux machines. This is only possible if files are checked-out `as-is` and line-breaks are always generated with the same EoL sequence (unix). If we use `autocrlf` a compilation on windows would yield different results than the compilation on unix. That's not desirable.

**Comment by LorenzoBettini** [comment]
As far as I know, not using `autocrlf=true` on Windows always leads to troubles; indeed, a few files ended up with the wrong EOL in the past (fixed by @HannesWell).

By removing the settings of Unix EOLs nothing broke.

**Comment by merks** [comment]
Start rant.

Here's what my experience says.  Tell everyone on Windows that they should use only Linux line endings.  Then if the tools are bad and generate only Linux line endings rather than respecting your settings, you won't notice, the tools will just remain broken, and the tools will never be fixed.   But that makes everyone happy, especially the people not on windows, so just do it.   😱 

End rant.

Note though that minimally you should use `autocrlf=input` because tools can be bad in multiple ways.  Some tools are bad that on Windows they always produce windows line endings because that's what the system property tells them to do.  As far as I can tell it's impossible to set the system property to a different value via a command line argument.  After all, who would ever want to do that, other than me?  So if you're not careful, your solution to the problem is exactly what will end up committing windows line endings into the repository causing your solution to be the new problem.

I can see many bad solutions in the project catalog:

![image](https://github.com/eclipse/xtext/assets/208716/d9fedad7-0cac-45cc-b40f-ffefa0b79fe2)
![image](https://github.com/eclipse/xtext/assets/208716/86b5c89e-538f-415b-8764-3758b22d4baa)

Here is a good one:

![image](https://github.com/eclipse/xtext/assets/208716/fa28e6da-b870-4060-ac64-fd6c8c3b4ca0)

**Comment by merks** [comment]
I'd leave it as true.  I wasn't sure where this thread was going, didn't have the time to read it in detail, but was hoping it would not lead to simply removing the setting because you can see how many other projects have reached that questionable conclusion.  Sorry for the noise.

**Comment by merks** [comment]
I think you've specified the task such that it always kicks in when you start.

![image](https://github.com/eclipse/xtext/assets/208716/71c8a7d6-aa75-43c6-88c7-2bfb4dd0a940)

Maybe you want it only for new projects added to the workspace?

**Comment by LorenzoBettini** [comment]
Mh.. that might be the cause, true! But the problem happens only in Windows, not in Linux, nor macOS...

**Comment by merks** [comment]
"The problem" being it always runs?

**Comment by LorenzoBettini** [comment]
Yes, exactly: when you restart Eclipse, it will always run; but only on Windows.

**Comment by merks** [comment]
I see.  That's quite unexpected!  I hate when OSes behave differently.  It's too much work to track down such problems...

**Comment by LorenzoBettini** [comment]
I'll check whether the Unix EOL settings were the culprit, when I'm on a Windows machine

**Comment by LorenzoBettini** [comment]
@Bananeweizen @pbrossel everything should be fixed now (together with https://github.com/eclipse/xtext/pull/3004)

Oomphing from scratch or updating with Oomph after updating to `main` should give you a clean workspace without dirty files on Windows.
IMPORTANT: on Windows you must have autocrlf=true

**Comment by szarnekow** [comment]
> IMPORTANT: on Windows you must have autocrlf=true

Is this already part of the oomph setup?

**Comment by LorenzoBettini** [comment]
> > IMPORTANT: on Windows you must have autocrlf=true
> 
> Is this already part of the oomph setup?

@szarnekow I don't think we can enforce that on a per-OS basis. Can we @merks ?

**Comment by merks** [comment]
Oomph's git clone task does this configuration by default for every clone unless you have an explicit configuration for that in the git clone task itself.   With some trickery I think it's possible do set it differently only different OSes, but given it's a no-op on Linux and Mac and I believe even there it would prevent checking in Windows line endings, so I don't see a need to configure it differently on different OSes.

**Comment by LorenzoBettini** [comment]
@merks so, if I understand correctly, by default Oomph sets autocrlf=true in Windows and does nothing in macOS and Linux, is that correct? If so, we're all fine.

**Comment by pbrossel** [comment]
> @Bananeweizen @pbrossel everything should be fixed now (together with #3004)
> 
> Oomphing from scratch or updating with Oomph after updating to `main` should give you a clean workspace without dirty files on Windows. IMPORTANT: on Windows you must have autocrlf=true

**Thanks a lot to all having worked on this issue !** 
Setup tasks are now finished quickly without rebuilding everything : 
![image](https://github.com/eclipse/xtext/assets/167133255/66a267fd-a129-4729-a013-bf9db8efacf4)

But there are some questions / remarks : 
- I moved to my fork by replacing the contents of `xtext-main\git\xtext` : **Is this the right approach ?** 
- I had to add `autocrlf = true` manually, maybe because I cloned the fork manually ?
- still had tons of fake changes even after resetting to `main`.  Many restarts or clean/rebuild did not help.
- target platform had errors. Editing and reapplying the definition helped to fix compile errors, but I got a warning that I should use a newer Eclipse version (currently using `Version: 2024-03 (4.31.0) Build id: 20240307-1437` )
- After reset hard + clean rebuild, I finally had 17 'fake' changes left over. **Is that OK ?**
```
org.eclipse.xtend.core.tests/xtend-gen/org/eclipse/xtend/core/tests/EmfModelsTest.java
org.eclipse.xtend.core/xtend-gen/org/eclipse/xtend/core/macro/ConstantExpressionsInterpreter.java
org.eclipse.xtend.core/xtend-gen/org/eclipse/xtend/core/macro/TransformationContextImpl.java
org.eclipse.xtend.core/xtend-gen/org/eclipse/xtend/core/macro/ValidationContextImpl.java
org.eclipse.xtend.core/xtend-gen/org/eclipse/xtend/core/macro/declaration/AnnotationReferenceBuildContextImpl.java
org.eclipse.xtend.core/xtend-gen/org/eclipse/xtend/core/macro/declaration/CompilationUnitImpl.java
org.eclipse.xtend.examples/projects/xtend-euler/xtend-gen/euler/Solution_004.java
org.eclipse.xtend.examples/projects/xtend-euler/xtend-gen/euler/Solution_014.java
org.eclipse.xtend.examples/projects/xtend-euler/xtend-gen/euler/Solution_017.java
org.eclipse.xtend.examples/projects/xtend-euler/xtend-gen/euler/Solution_021.java
org.eclipse.xtend.examples/projects/xtend-examples/xtend-gen/example2/BasicExpressions.java
org.eclipse.xtend.ide.swtbot.tests/swtbot/xtend-gen/org/eclipse/xtend/ide/tests/SwtBotProjectHelper.java
org.eclipse.xtend.ide/xtend-gen/org/eclipse/xtend/ide/codebuilder/AbstractCodeBuilder.java
org.eclipse.xtend.lib/xtend-gen/org/eclipse/xtend/lib/annotations/DelegateProcessor.java
org.eclipse.xtend.lib/xtend-gen/org/eclipse/xtend/lib/annotations/EqualsHashCodeProcessor.java
org.eclipse.xtext.xbase/xtend-gen/org/eclipse/xtext/xbase/compiler/JvmModelGenerator.java
org.eclipse.xtext.xbase/xtend-gen/org/eclipse/xtext/xbase/controlflow/ConstantConditionsInterpreter.java
```
- **is there a documentation somewhere about the steps to achieve to catch up with a new change or milestone ?** 
- when starting the ide, there is still something building, even if no changes occurred. Takes some time but I can live with it. **Is that OK ?**

**Comment by LorenzoBettini** [comment]
@pbrossel Very quick answer about the crucial part (I'll try to answer the other questions ASAP): if you clone a repository manually, core.autocrlf=true must be set appropriately in the user global git configuration (~/.gitconfig), and, IIRC, it must be done before cloning. Concerning the fake changes, are they really fake or do they still contain differences? Note that you must have the latest Xtend in your IDE so that you have the version with these changes https://github.com/eclipse/xtext/pull/3004.

**Comment by pbrossel** [comment]
@LorenzoBettini 

> Very quick answer about the crucial part (I'll try to answer the other questions ASAP): if you clone a repository manually, core.autocrlf=true must be set appropriately in the user global git configuration (~/.gitconfig), and, IIRC, it must be done before cloning. 

I'll retry it according to your advice.

>Concerning the fake changes, are they really fake or do they still contain differences? 

I thought they were 'fake' changes, but it could be the result of using the wrong version of xtend in the ide.

![image](https://github.com/eclipse/xtext/assets/167133255/062313ca-e8fa-468e-b94d-29c8d5edc1fb)

At least these 4 generated sources come back to the original version after installing the IDE updates.

![image](https://github.com/eclipse/xtext/assets/167133255/c586fb2c-2877-453e-8900-b10f060db42e)


>Note that you must have the latest Xtend in your IDE so that you have the version with these changes #3004.

Where can I check if the changes are contained in the updates ?  
![image](https://github.com/eclipse/xtext/assets/167133255/12f3d9e1-2baa-4a03-82b0-e5640964f603)

**Comment by LorenzoBettini** [comment]
@pbrossel your version of Xtend is the latest one, which contains https://github.com/eclipse/xtext/pull/2993, where @HannesWell replaced the use of Guava `Object.equal` with Java `Object.equals`.

After you install the new version of the Xtend compiler, as you have already done, you have to trigger regeneration in the projects showing dirty files (since now the full build clean Oomph task is not triggered anymore on existing projects, as you saw above). There's no need to fully clean the whole workspace: only cleaning the single projects with dirty Java files is enough.

Please, let me know.

**Comment by pbrossel** [comment]
> After you install the new version of the Xtend compiler, as you have already done, you have to trigger regeneration in the projects showing dirty files (since now the full build clean Oomph task is not triggered anymore on existing projects, as you saw above). There's no need to fully clean the whole workspace: only cleaning the single projects with dirty Java files is enough.
> 
> Please, let me know.

I first removed the 'fake' changes,  and rebased on your latest commit and got no changes at all : ok.

![image](https://github.com/eclipse/xtext/assets/167133255/47ac9516-9789-45f0-a0d6-c174535bf86e)

I then cleaned and rebuilt the first project in the list of fake changes `org.eclipse.xtend.core` and got changes in these 8 files :

![image](https://github.com/eclipse/xtext/assets/167133255/aec45782-ec6c-49b9-a5b5-4a54bd63e148)

Opening the compare editor by double clicking on the first unstaged change shows `no difference` on the java file.

![image](https://github.com/eclipse/xtext/assets/167133255/804c3fe3-3968-428f-8450-dc35c292ef46)

Opening the working tree version using the context menu on the first unstaged change opens the xtend source. 
Comparing with the latest commit shows `'1 difference'` being something like a fake (i.e. the button `next difference` does not show anything).

![image](https://github.com/eclipse/xtext/assets/167133255/a818e495-4f95-48d6-9a60-5af2dab1c876)

The button `next change` shows indeed that something is wrong : 

![image](https://github.com/eclipse/xtext/assets/167133255/75426dc4-1cbd-46d9-b9a6-9ade9d401aea)

So I guess that there are still some fake changes around.

I get 219 changes with the same symptoms after having done a full clean rebuild :

![image](https://github.com/eclipse/xtext/assets/167133255/1757589d-5f2e-460b-8ae9-60a4062a4f90)


**Note that I did not clone again after setting user global git configuration to autocrlf=true** 

I'll try to start from scratch with a fresh clone and see if it solves the problem.

**Comment by pbrossel** [comment]
@LorenzoBettini 

Even if setting the core.autocrlf=true in the user settings before cloning (via Egit), the clone's config does not contain the setting.

![image](https://github.com/eclipse/xtext/assets/167133255/01e2f241-037f-47ca-864e-92d47c312bb2)

 I now added the `core.autocrlf=true` manually just to make sure it stays good on repo level.

**The new clone works better !** 👍 but a full rebuild still shows  2 dirty files : 

![image](https://github.com/eclipse/xtext/assets/167133255/342cbd6b-715a-49fb-909a-ee03c2491b42)

![image](https://github.com/eclipse/xtext/assets/167133255/eb29ed5a-672f-41df-8868-5d4e0cc91d72)

You can see the details of the changes [here](https://github.com/eclipse/xtext/commit/cfcfd539660fb4743b9f33a179489d5936f200c7)

**Comment by cdietrich** [comment]
this seems to be the inverse change of
https://github.com/eclipse/xtext/pull/3022/files
maybe something got fixed in jdt again

**Comment by pbrossel** [comment]
@cdietrich 

> which target platform did you select in oomoph? xtext-latest/2024-06?

If I remember well, I selected 'latest' when installing 2 weeks or so ago : 2024-03 was the `latest` at that time.

![image](https://github.com/eclipse/xtext/assets/167133255/e5f69155-c9b0-4deb-97c3-79efb746fe73)

The currently starting IDE still shows `2024-03`.

![image](https://github.com/eclipse/xtext/assets/167133255/92093c33-425a-4e92-8dad-b60e69902e03)

This is the target platform setting : 

![image](https://github.com/eclipse/xtext/assets/167133255/f752218c-19f5-468e-9ac4-c13731164889)

But `/org.eclipse.xtext.contributor/Xtext.setup` shows `2024-06` which seems to be right.

![image](https://github.com/eclipse/xtext/assets/167133255/2a4ee4c0-03be-43e2-abc0-a3afcff890d4)

**I guess I have to do the setup again, or can I ask oopmh to do it ?**

**Comment by merks** [comment]
Try Help -> Perform Setup Tasks.  That should update everything, including the installation.

**Comment by pbrossel** [comment]
Thanks for the hint.
I ran it, it did a couple of updates, but I still see `2024-03` when starting the IDE and the installation details still show : 

![image](https://github.com/eclipse/xtext/assets/167133255/31422161-66ee-45f6-a835-d4e70969fd8c)

Do I have to switch the active target platform to `xtext-latest` instead of `Modular Target`? 

![image](https://github.com/eclipse/xtext/assets/167133255/76ee2c52-ffc9-40f0-9776-ad4fcffb7817)

**Comment by LorenzoBettini** [comment]
The version of your eclipse and the version of the target platform are two different things. The version of eclipse is the one you selected when you select the distribution in oomph, the target platform is the one you select as an xtext property. The target platform is used for compiling and running tests.

**Comment by merks** [comment]
You guys need an oomph configuration that chooses the latest product version so that things always stay up to date.  Open an issue and I will help. But I have no time this week.

**Comment by pbrossel** [comment]
@LorenzoBettini Yes, I agree. But I got 2 dirty files when doing a full rebuild i.e. compiling because of having the wrong eclipse version (if I understood @cdietrich right : https://github.com/eclipse/xtext/issues/2778#issuecomment-2094643170).

It's not a problem for the issue I'm currently working on https://github.com/eclipse/xtext/issues/2337#issuecomment-2057693365.

Not easy for newcomers like me to guess the steps to do to catch up with the latest changes in the project. 
Many thanks to all for the help.

**Comment by LorenzoBettini** [comment]
@merks do you mean an issue here in Xtext?

@pbrossel I think @cdietrich referred to the target platform, which you can update either with Oomph by selecting the latest one or by opening the `xtext-latest.target` in Eclipse, wait for resolution, and set it is as current (or choose reload).
Concerning the running IDE, IIRC, you have to choose either the "Latest" (currently 2024-06) or "Latest Released" (currently 2024-03). Once you choose that, you cannot switch it: upon updating, it will update your Eclipse according to the stable (latest released) or the current milestone (latest). Is that right, @merks ? If you want to switch, you have to run Oomph from scratch.

Concerning the steps to catch up with the latest changes, I don't think there's a real workflow but to update now and then. ;) Now and then, "dirty" files like the two you mentioned can show up. Since you already have the latest Xtend in Eclipse, it should be enough to update the target platform. I had never seen the warning from PDE. You might try to ignore that.

**Comment by merks** [comment]
You can use Navigate -> Open Setup -> Installation and via the properties view can switch to any product version:

![image](https://github.com/eclipse/xtext/assets/208716/ec1d7f1e-00f5-4736-a47c-c80095e3bf67)

Other projects have a configuration:

- https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md#creating-an-eclipse-development-environment
- https://github.com/eclipse-simrel
- https://github.com/eclipse-birt/birt#create-a-birt-development-environment

This allows you to determine/specify exactly which product and which version is to be installed, and can specify variable choices so the user can't make mistakes and the user needs to read fewer instructions.

It's quite simple:

https://raw.githubusercontent.com/eclipse/birt/master/build/org.eclipse.birt.releng/BIRTConfiguration.setup

**Comment by cdietrich** [comment]
i still wonder what jdt change causes the last problem.

**Comment by pbrossel** [comment]
@merks I changed the IDE version successfully. Thanks for the advice. 
This setup has to be executed manually as soon as the next 'latest' comes out, right ?

![image](https://github.com/eclipse/xtext/assets/167133255/93fb7b68-aa32-4a9e-8355-e2448859ea55)

Should I change the `jdk `settings too ? I had to replace it because of 'Failed to init ct.sym' errors, but do not remember where I changed it.

![image](https://github.com/eclipse/xtext/assets/167133255/0652c0af-4984-4676-8eaa-ed057bf4df3c)

Oomph just allows to select a jre, not a jdk.

![image](https://github.com/eclipse/xtext/assets/167133255/01d9d1e2-c6a4-4dc2-a248-9dcd4f4c4353)

The whole discussion was very interesting and I learned a lot.
Automating the setup would nevertheless make life easier...

@cdietrich I can confirm that the 2 dirty files have disappeared after a clean rebuild. 

@LorenzoBettini Did you find some time to look into this ?
>I moved to my fork by replacing the contents of xtext-main\git\xtext : **Is this the right approach** ?

**Comment by LorenzoBettini** [comment]
> @merks I changed the IDE version successfully. Thanks for the advice. This setup has to be executed manually as soon as the next 'latest' comes out, right ?

If you perform setup tasks (not from scratch), it will update your IDE.
In general, you must perform setup tasks if you want to stay up to date.

A small warning: using the "latest" in your IDE (instead of "latest released") might expose you to some bugs introduced in the currently developed Eclipse or one of its plugins (e.g., https://github.com/eclipse/xtext/issues/3018).


> Oomph just allows to select a jre, not a jdk.
> 

You can manually add a JDK to let Oomph know about that.
IIRC, it will automatically detect system JDK installations.

> 
> The whole discussion was very interesting and I learned a lot. Automating the setup would nevertheless make life easier...

well... it's automated... it's the small initial configuration that requires a few steps ;)

> 
> @cdietrich I can confirm that the 2 dirty files have disappeared after a clean rebuild.

Great! :)

> 
> @LorenzoBettini Did you find some time to look into this ?
> 
> > I moved to my fork by replacing the contents of xtext-main\git\xtext : **Is this the right approach** ?

sorry, I had missed that.

Your solution is a bit... "brutal"... ;)

in general, not specific of Xtext, you first fork the github project, and then, during the Oomph setup, you specify the URL of your fork, instead of the main one.

Alternatively, if you cloned the main one, you can add your fork as an additional Git remote (either from the command line or from Egit); of course, no need to clone it from scratch, just as a remote. Of course, when you push, you have to push to the remote of your fork.

In general, it's good to have also the main repository as a remote in order to easily keep your fork up to date with the main one.

**Comment by pbrossel** [comment]
@LorenzoBettini Thanks a lot! All open questions got answered 👍 !
I will follow your advice.
